### PR TITLE
feat(animations): support function types in transitions

### LINKED ISSUE #13537: There is no way to disable animations for transitions without dynamic states
```
// disable here
transition('* => *', [])

// allow here
transition('* => *', [ ... ])
```

It would be nice to have this support a function to determine which transition to use dynamically.

### PR DESCRIPTION
**Pull request by matsko**
Closes #13537

### CODE REVIEW

**Review by mhevery — APPROVED** [review]

File: modules/@angular/compiler-cli/integrationtest/src/animate.ts [hunk 1]
```diff
@@ -16,7 +20,7 @@ import {AUTO_STYLE, Component, animate, state, style, transition, trigger} from
         state('*', style({height: AUTO_STYLE, color: 'black', borderColor: 'black'})),
         state('closed, void', style({height: '0px', color: 'maroon', borderColor: 'maroon'})),
         state('open', style({height: AUTO_STYLE, borderColor: 'green', color: 'green'})),
-        transition('* => *', animate(500))
```
Comment (tbosch) [inline modules/@angular/compiler-cli/integrationtest/src/animate.ts]: Could you keep the old transition and add another one?

File: modules/@angular/compiler-cli/integrationtest/src/animate.ts [hunk 2]
```diff
@@ -8,6 +8,10 @@
 
 import {AUTO_STYLE, Component, animate, state, style, transition, trigger} from '@angular/core';
 
+export function anyToAny(stateA: string, stateB: string): boolean {
+  return true;
```
Comment (tbosch) [inline modules/@angular/compiler-cli/integrationtest/src/animate.ts]: Could you add a simple condition in here?

File: modules/@angular/compiler/src/animation/animation_ast.ts [hunk 3]
```diff
@@ -49,6 +49,10 @@ export class AnimationStateTransitionExpression {
   constructor(public fromState: string, public toState: string) {}
 }
 
+export class AnimationStateTransitionFnExpression extends AnimationStateTransitionExpression {
+  constructor(public fn: Function) { super(null, null); }
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_ast.ts]: The correct type is `Function | StaticSymbol`

File: modules/@angular/compiler/src/animation/animation_parser.ts [hunk 4]
```diff
@@ -110,9 +110,12 @@ function _parseAnimationStateTransition(
     errors: AnimationParseError[]): AnimationStateTransitionAst {
   const styles = new StylesCollection();
   const transitionExprs: AnimationStateTransitionExpression[] = [];
-  const transitionStates = transitionStateMetadata.stateChangeExpr.split(/\s*,\s*/);
+  const stateChangeExpr = transitionStateMetadata.stateChangeExpr;
+  const transitionStates: Array<Function|string> = typeof stateChangeExpr == 'string' ?
+      (<string>stateChangeExpr).split(/\s*,\s*/) :
+      [<Function>stateChangeExpr];
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_parser.ts]: Type cast: `<Function|StaticSymbol>`

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 5]
```diff
@@ -39,7 +39,11 @@ export class CompileAnimationStateDeclarationMetadata extends CompileAnimationSt
 }
 
 export class CompileAnimationStateTransitionMetadata extends CompileAnimationStateMetadata {
-  constructor(public stateChangeExpr: string, public steps: CompileAnimationMetadata) { super(); }
+  constructor(
+      public stateChangeExpr: string|((stateA: string, stateB: string) => boolean),
```
Comment (tbosch) [inline modules/@angular/compiler/src/compile_metadata.ts]: Missing `|StaticSymbol`

File: modules/@angular/compiler-cli/integrationtest/src/animate.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler-cli/integrationtest/src/animate.ts]: I.e. just make another `transition`?

**Pull request closed without merging**